core/service/manager.go

```diff
@@ -78,49 +79,75 @@ type Manager struct {
 	discoveryFactory DiscoveryFactory
 }
 
-// Start starts an instance of the given service type if knows one in service registry. The method blocks.
+// Start starts an instance of the given service type if knows one in service registry.
 // It passes the options to the start method of the service.
 // If an error occurs in the underlying service, the error is then returned.
-func (manager *Manager) Start(providerID identity.Identity, serviceType string, options Options) (err error) {
+func (manager *Manager) Start(providerID identity.Identity, serviceType string, options Options) (id ID, err error) {
 	service, proposal, err := manager.serviceRegistry.Create(serviceType, options)
 	if err != nil {
-		return err
+		return id, err
 	}
 
 	dialogWaiter, err := manager.dialogWaiterFactory(providerID, proposal.ServiceType)
 	if err != nil {
-		return err
+		return id, err
 	}
 	providerContact, err := dialogWaiter.Start()
 	if err != nil {
-		return err
+		return id, err
 	}
 	proposal.SetProviderContact(providerID, providerContact)
 
 	dialogHandler := manager.dialogHandlerFactory(proposal, service)
 	if err = dialogWaiter.ServeDialogs(dialogHandler); err != nil {
-		return err
+		return id, err
 	}
 
 	discovery := manager.discoveryFactory()
 	discovery.Start(providerID, proposal)
 
-	err = service.Serve(providerID)
-	if err != nil {
-		return err
-	}
-	manager.servicePool.Add(&Instance{
+	instance := Instance{
+		state:        Starting,
 		service:      service,
 		proposal:     proposal,
 		dialogWaiter: dialogWaiter,
 		discovery:    discovery,
-	})
+	}
+
+	id, err = manager.servicePool.Add(&instance)
+	if err != nil {
+		return id, err
+	}
 
-	discovery.Wait()
-	return nil
+	go func() {
+		instance.state = Running
+		err = service.Serve(providerID)
+		instance.state = NotRunning
```

Comment:
State  is handled very strange here, why it's becoming to `NotRunning` immediately

Reply:
`Serve` is blocking operation, the state became `NotRunning` only when service finished to `Serve`.